lib/blamer/message.rb

```diff
@@ -0,0 +1,200 @@
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+module Blamer
+
+  # This blamer uses the recency of the commit that most recently modified a
+  # given stack trace element to determine how likely that line is to be at
+  # fault.
+  #
+  # Matching to a Bug
+  # -----------------
+  #
+  # In order to match an Occurrence to a Bug, the line number of the backtrace
+  # that is at fault must first be determined. This is done probabilistically, by
+  # assigning a score to each line of the backtrace and picking the line with
+  # the highest score. Currently, the factors that influence a line's score are
+  #
+  # * the "height" of that line in the backtrace, and
+  # * the age of the commit that most recently modified that line.
+  #
+  # The highest-scoring file and line number are termed the **relevant file** and
+  # **relevant line**.
+  #
+  # An Occurrence is matched to a Bug using three mandatory criteria:
+  #
+  # * the relevant file path,
+  # * the relevant line number, and
+  # * the class name of the raised exception.
+  #
+  # In addition, if a "blamed" commit is available for the relevant file and line,
+  # the bug is matched on that commit ID as well. Note that the exception message
+  # is _not_ included in the search criteria.
+  #
+  # While each Occurrence maintains the original exception message, the Bug
+  # filters the exception message to remove occurrence-specific information. There
+  # are two ways of filtering exception messages:
+  #
+  # * If the exception can be filtered using {MessageTemplateMatcher}, that
+  #   filtered message is used.
+  # * Otherwise, a general search-and-replace is used to replace likely
+  #   occurrence-specific information (like numbers, inspect values, etc.) with
+  #   placeholders.
+  #
+  # All placeholders take the form "[ALL CAPS IN BRACKETS]" so that the view can
+  # recognize them as such and format them appropriately.
+
+  class Message < Base
+    def find_or_create_bug!
+      
+      criteria = bug_search_criteria
+      bug      = if deploy
+                   # for distributed projects, we need to find either
+                   # * a bug associated with the exact deploy specified in the build attribute
+                   # * or, if the exact deploy specified does not have such a bug:
+                   #   * find any open (not fixed) bug matching the criteria;
+                   #   * if such a bug exists, use that bug and alter its deploy_id to
+                   #     the given deploy;
+                   #   * otherwise, a new bug is created.
+                   b        = Bug.transaction do
```

Comment:
Tell me more about this indentation style. Is it just github mal-formatting it here?